Exception specification mismatch with newer glibc for host math functions

Newer versions of glibc (2.41+?) declare `rsqrt`, `sinpi` and `cospi` (and their float variants) as `noexcept(true)`. This conflicts with the definitions in chipStar and leads to compile errors like:

```
In file included from /usr/include/math.h:467:
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:83:1: error: exception specification in declaration does not match previous declaration
   83 | __MATHCALL_VEC (cospi,, (_Mdouble_ __x));
```

This can be fixed by adding `noexcept` to the corresponding function definitions in `include/hip/devicelib/host_math_funcs.hh`.

Additionally, the functions in glibc have C linkage, so this needs to be matched for `sinpif`, `cospif`, `rsqrt` and `rsqrtf`. 

Both changes should be fine also with older glibc, but I haven't tested it.

--- a/include/hip/devicelib/host_math_funcs.hh
+++ b/include/hip/devicelib/host_math_funcs.hh
@@ -26,11 +26,11 @@ extern "C" {
 }
 
 // Trigonometric functions
-inline float cospi(float x) { return __ocml_cospi_f32(x); }
-inline float cospif(float x) { return __ocml_cospi_f32(x); }
+inline float cospi(float x) noexcept { return __ocml_cospi_f32(x); }
+extern "C" inline float cospif(float x) noexcept { return __ocml_cospi_f32(x); }
 
-inline float sinpi(float x) { return __ocml_sinpi_f32(x); }
-inline float sinpif(float x) { return __ocml_sinpi_f32(x); }
+inline float sinpi(float x) noexcept { return __ocml_sinpi_f32(x); }
+extern "C" inline float sinpif(float x) noexcept { return __ocml_sinpi_f32(x); }
 
 // Error functions
 inline double erfcinv(double x) { return __ocml_erfcinv_f64(x); }
@@ -71,9 +71,9 @@ inline void sincospif(float x, float* psin, float* pcos) {
 // Sign bit
 inline int signbit(double x) { return __ocml_signbit_f64(x); }
 
-inline double rsqrt(double x) { return __ocml_rsqrt_f64(x); }
+extern "C" inline double rsqrt(double x) noexcept { return __ocml_rsqrt_f64(x); }
 
-inline float rsqrtf(float x) { return __ocml_rsqrt_f32(x); }
+extern "C" inline float rsqrtf(float x) noexcept { return __ocml_rsqrt_f32(x); }
 
 
 #endif // HIP_INCLUDE_DEVICELIB_HOST_MATH_FUNCS_H

